File: Algorithms/Python/0022/0022.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []
        def DFS(nLeft: int, nRight: int, t: str):
            if nLeft == n and nRight == n:
                res.append(t)
                return
            if nLeft < n:
                DFS(nLeft + 1, nRight, t + "(")
            if nRight < nLeft:
                DFS(nLeft, nRight + 1, t + ")")
        DFS(0, 0, "")
        return res

# class Solution:
#     def generateParenthesis(self, n: int) -> List[str]:
#         if n == 0:
#             return [""]
#         res = []
#         for i in range(n):
#             for

File: Algorithms/C++/0022/0022.cpp
class Solution {
public:
    vector<string> res;
    vector<string> generateParenthesis(int n) {
        DFS(0, 0, n, "");
        return res;
    }
    void DFS(int nleft, int nright, int n, string t) {
        if(nleft == n && nright == n){
            res.push_back(t);
            return;
        }
        if (nleft < n) {
            DFS(nleft + 1, nright, n, t + "(");
        }
        if (nright < n) {
            if(nright < nleft) {
                DFS(nleft, nright + 1, n, t + ")");
            }
        }
    }
};
